samza-kafka/src/main/java/org/apache/samza/system/kafka/KafkaSystemConsumer.java

```diff
@@ -330,6 +320,73 @@ public String getSystemName() {
     return systemName;
   }
 
+  @VisibleForTesting
+  static class KafkaStartpointRegistrationHandler implements StartpointVisitor {
+
+    private final Consumer kafkaConsumer;
+
+    KafkaStartpointRegistrationHandler(Consumer kafkaConsumer) {
+      this.kafkaConsumer = kafkaConsumer;
+    }
+
+    @Override
+    public void visit(SystemStreamPartition systemStreamPartition, StartpointSpecific startpointSpecific) {
+      TopicPartition topicPartition = toTopicPartition(systemStreamPartition);
+      long offsetInStartpoint = Long.parseLong(startpointSpecific.getSpecificOffset());
+      LOG.info("Updating the consumer fetch offsets of topic partition: {} to {}.", topicPartition, offsetInStartpoint);
+
+      // KafkaConsumer is not thread-safe.
+      synchronized (kafkaConsumer) {
+        kafkaConsumer.seek(topicPartition, offsetInStartpoint);
+      }
+    }
+
+    @Override
+    public void visit(SystemStreamPartition systemStreamPartition, StartpointTimestamp startpointTimestamp) {
+      Long timestampInStartpoint = startpointTimestamp.getTimestampOffset();
+      TopicPartition topicPartition = toTopicPartition(systemStreamPartition);
+      Map<TopicPartition, Long> topicPartitionsToTimeStamps = ImmutableMap.of(topicPartition, timestampInStartpoint);
+
+      // Look up the offset by timestamp.
+      LOG.info("Looking up the offsets of the topic partition: {} by timestamp: {}.", topicPartition, timestampInStartpoint);
+      Map<TopicPartition, OffsetAndTimestamp> topicPartitionToOffsetTimestamps = kafkaConsumer.offsetsForTimes(topicPartitionsToTimeStamps);
```

Comment:
According to the [Kafka documentation](https://kafka.apache.org/0110/javadoc/org/apache/kafka/clients/consumer/KafkaConsumer.html#offsetsForTimes(java.util.Map)), `OffsetAndTimestamp` may be null for a partition if no events exist at or after the specified timestamp. Probably need to handle null in this case. Seeking to `upcoming` probably makes the most sense.

Reply:
Good question. 
Excerpt the kafka javadoc linked in the above comment:
```
If the message format version in a partition is before 0.10.0, i.e. the messages do not have timestamps, null will be returned for that partition
```

Null would be returned by  KafkaConsumer `offsetsForTimes` API when used with the kafka broker version < `0.10.0`. 

Starting from 1.0, samza supports kafka version: `0.11.0.2`. From the [kafka-compatibility-matrix](https://cwiki.apache.org/confluence/display/KAFKA/Compatibility+Matrix), kafka-client version used by samza is not compatible with broker versions < `0.10.0`. 

Since samza doesn't support kafka brokers < 0.10.0, I'm not sure if we should handle null value returned by this API.